src/ProjectSerializer.cpp

```diff
@@ -89,115 +94,261 @@ XO("This recovery file was saved by Audacity 2.3.0 or before.\n"
+static const auto ReadULong =
+   IsLittleEndian() ? &ReadLittleEndian<ULong> : &ReadBigEndian<ULong>;
+static const auto ReadLongLong =
+   IsLittleEndian() ? &ReadLittleEndian<LongLong> : &ReadBigEndian<LongLong>;
+
+// Functions to read and write certain lengths -- maybe we will change
+// our choices for widths or signedness?
+
+using Length = Int; // Instead, as wide as size_t?
+static const auto WriteLength = WriteInt;
+static const auto ReadLength = ReadInt;
+
+using Digits = Int; // Instead, just an unsigned char?
+static const auto WriteDigits = WriteInt;
+static const auto ReadDigits = ReadInt;
+
+class XMLTagHandlerAdapter final
+{
+public:
+   explicit XMLTagHandlerAdapter(XMLTagHandler* handler) noexcept
+       : mBaseHandler(handler)
+   {
+   }
+
+   void EmitStartTag(const std::string_view& name)
+   {
+      if (mInTag)
+         EmitStartTag();
+
+      mCurrentTagName = name;
+      mInTag = true;
+   }
+
+   void EndTag(const std::string_view& name)
+   {
+      if (mInTag)
+         EmitStartTag();
+
+      if (XMLTagHandler* const handler = mHandlers.back())
+         handler->HandleXMLEndTag(name);
+
+      mHandlers.pop_back();
+   }
+
+   void WriteAttr(const std::string_view& name, std::string value)
+   {
+      assert(mInTag);
+
+      if (!mInTag)
+         return;
+
+      mAttributes.emplace_back(name, CacheString(std::move(value)));
+   }
 
-   using UShort = std::uint16_t;
-   using Int = std::int32_t;
+   template <typename T> void WriteAttr(const std::string_view& name, T value)
+   {
+      assert(mInTag);
 
-   using Long = std::int32_t;   // To save long values
-   using ULong = std::uint32_t; // To save size_t values
+      if (!mInTag)
+         return;
 
-   using LongLong = std::int64_t;
+      mAttributes.emplace_back(name, XMLAttributeValueView(value));
+   }
 
-   // Detect this computer's endianness
-   bool IsLittleEndian()
+   void WriteData(std::string value)
    {
-      const std::uint32_t x = 1u;
-      return
-         static_cast<const unsigned char*>(static_cast<const void*>(&x))[0];
-      // We will assume the same for other widths!
+      if (mInTag)
+         EmitStartTag();
+
+      if (XMLTagHandler* const handler = mHandlers.back())
+         handler->HandleXMLContent(CacheString(std::move(value)));
    }
-   // In C++20 this could be
-   // constexpr bool IsLittleEndian = (std::endian::native == std::endian::little);
-   // static_assert( IsLittleEndian || (std::endian::native == std::endian::big),
-   //    "Oh no!  I'm mixed-endian!" );
-
-   // Functions that can read and write native integer types to a canonicalized
-   // little-endian file format.  (We don't bother to do the same for floating
-   // point numbers.)
-
-   // Write native little-endian to little-endian file format
-   template< typename Number >
-   void WriteLittleEndian( MemoryStream &out, Number value )
+
+   void WriteRaw(std::string value)
    {
-      out.AppendData( &value, sizeof(value) );
```

Comment:
I thought the emptiness of this function was a bug until I stepped in the debugger and figured out that it is visited only when reading back the xml header data from  ProjectFileIO::WriteXMLHeader which we don't really need.  (But still write to avoid compatibility breaks, right?)

A comment explaining that would help.